### PR TITLE
[WIP] Job template for content source change

### PR DESCRIPTION
**Pull request by stejskalleos**
Waiting for https://github.com/Katello/katello/pull/9780/files

### CODE REVIEW

**Comment by stejskalleos** [comment]
@ares @ehelms @evgeni

File: app/views/foreman/job_templates/change_content_source.erb [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+test -f $RHSM_CFG.bak || cp $RHSM_CFG $RHSM_CFG.bak
+
+subscription-manager config \
+  --server.hostname="<%= rhsm_url.host %>" \
+  --server.port="<%= rhsm_url.port %>" \
+  --server.prefix="<%= rhsm_url.path %>" \
+  --rhsm.repo_ca_cert="$KATELLO_SERVER_CA_CERT" \
+  --rhsm.baseurl="<%= pulp_content_url %>"
+
+# Update host
```
Comment (ekohl) [inline app/views/foreman/job_templates/change_content_source.erb:78]: Why is this needed? I would expect the workflow to be:
* User changes content source on the host
* Runs REX job to match the config on the host to what's in Foreman

Essentially: treat Foreman's DB as the source of truth.

Comment (ehelms) [reply app/views/foreman/job_templates/change_content_source.erb:78]: The primary idea is as you call out to make use of REX to enable changing the proxy a host is pointed at. The secondary idea is to use this same mechanism to allow changing the proxy from the host itself as users do with `bootstrap.py` since we have not achieved REX everywhere across installations. The workflow would be:

 * User on a host hits API to change source
 * Template is rendered and delivered to the host and ran
 
I imagine this being used by an Ansible module or a CLI tool like the newer RHC RHEL is working to provide.

File: app/models/katello/concerns/smart_proxy_extensions.rb [hunk 2]
```diff
@@ -444,11 +444,19 @@ def smart_proxy_service
 
       def rhsm_url
         if pulp_primary?
-          "https://#{URI.parse(url).host}/rhsm"
+          URI("https://#{URI.parse(url).host}/rhsm")
```
[the remark this answers is not in the document]
Comment (stejskalleos) [reply app/models/katello/concerns/smart_proxy_extensions.rb:447]: On my katello-devel-setup
```
proxy = SmartProxy.find(1)

proxy.setting(SmartProxy::PULP3_FEATURE, 'rhsm_url')
# => "https://localhost/rhsm"

proxy.rhsm_url
# => "https://centos7-katello-devel-stable.example.com/rhsm"
```

Is the value of `rhsm_url` in proxy plugin set by installer or users have to do it manually?
Because if they have to set it manually, there is chance that they didn't do it and it will break the workflow.

[the remark this answers is not in the document]
Comment (ekohl) [reply app/models/katello/concerns/smart_proxy_extensions.rb:447]: It will be set by the installer (https://github.com/theforeman/puppet-foreman_proxy_content/pull/390) but that hasn't been merged yet. You will also need a fallback for n-1 support. Foreman Proxy 3.0 won't have it and it's supported to register a 3.0 Proxy with a 3.1 Foreman. This fallback is the current code.

File: app/views/foreman/job_templates/change_content_source.erb (hunk 1 above)
Comment (ekohl) [reply app/views/foreman/job_templates/change_content_source.erb:78]: > The primary idea is as you call out to make use of REX to enable changing the proxy a host is pointed at.

I'd be against that if I'm honest. I've always seen Foreman as the source of truth. To me it would make more sense to have a "trigger" that if you change the content source on a host that a REX job can automatically be triggered to update the host.

I also see security concerns. Curl runs on the host and there the username and password will be visible in the process tree (`ps aux`). I hadn't mentioned it inline since I first want to have the higher level discussion.



> The workflow would be:
> 
>     * User on a host hits API to change source
> 
>     * Template is rendered and delivered to the host and ran

This makes sense to me and what I was arguing for: treat Foreman as the source of truth and then rectify hosts to match.

File: app/models/katello/concerns/smart_proxy_extensions.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (wbclark) [reply app/models/katello/concerns/smart_proxy_extensions.rb:447]: It's set by the installer as of https://github.com/theforeman/puppet-foreman_proxy_content/pull/384 but the apache vhost to proxy /rhsm on 443 is deployed in https://github.com/theforeman/puppet-foreman_proxy_content/pull/390

File: app/views/foreman/job_templates/change_content_source.erb (hunk 1 above)
Comment (stejskalleos) [reply app/views/foreman/job_templates/change_content_source.erb:78]: > * User on a host hits API to change source
> * Template is rendered and delivered to the host and ran

That was original plan, and done in https://github.com/Katello/katello/pull/9713, but then it was decided to go with REX template - it's more customizable and there is no need to maintain another API endpoint

Comment (ekohl) [reply app/views/foreman/job_templates/change_content_source.erb:78]: In that PR I also see a HTTP call to Foreman to modify data. That is still my problem with this. I think this REX template should be reduced to reading the host property and applying that info to the host in practice. So essentially this REX template without any job parameters and without this curl request.

Comment (stejskalleos) [reply app/views/foreman/job_templates/change_content_source.erb:78]: > I also see security concerns.

We could fix it fairly easy. Create new template macro `user_api_auth_token`, that would generate API auth token for the current user. Of course it would allow to set the scope and expiration time, same way as we do in the registration template [0]. 
This way we can call Foreman API without exposing user's credentials. Thoughts?

[0] https://github.com/theforeman/foreman/blob/develop/app/controllers/concerns/foreman/controller/registration_commands.rb#L39

Comment (ekohl) [reply app/views/foreman/job_templates/change_content_source.erb:78]: It still leaves a window of attack and I don't see the need to. Going Foreman -> Host -> Foreman feels plain wrong to me.

File: app/models/katello/concerns/smart_proxy_extensions.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (ekohl) [reply app/models/katello/concerns/smart_proxy_extensions.rb:447]: Since it should land in 4.3 I opened https://github.com/Katello/katello/pull/9777 to implement this.

File: app/views/foreman/job_templates/change_content_source.erb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+description_format: Configure subscription manager to new content source (smart proxy)
+feature: katello_change_content_source
+template_inputs:
+- name: smart_proxy
+  required: true
+  input_type: user
+  value_type: resource
+  resource_type: SmartProxy
+  description: New content source (Smart Proxy)
+  advanced: false
```
Comment (ekohl) [inline app/views/foreman/job_templates/change_content_source.erb:16]: IMHO this template should not have any inputs. It should use `host.content_source` as the Smart Proxy.

**Review by ekohl — CHANGES_REQUESTED** [review]

File: app/views/foreman/job_templates/change_content_source.erb (hunk 3 above)
Comment (stejskalleos) [reply app/views/foreman/job_templates/change_content_source.erb:16]: That would not work in case of rendering `app/views/foreman/job_templates/change_content_source_and_host.erb` template.

When `change_content_source_and_host.erb` template is rendered, `host.content_source` is still the old one, while `smart_proxy` input is the new one.

So running `change_content_source_and_host` job would not configure `subscription-manager` to the new proxy but to the old one.

Comment (ekohl) [reply app/views/foreman/job_templates/change_content_source.erb:16]: > That would not work in case of rendering `app/views/foreman/job_templates/change_content_source_and_host.erb` template.

Then I'd recommend to extracted it to a snippet that is properly reusable.

File: app/models/katello/concerns/smart_proxy_extensions.rb [hunk 4]
```diff
@@ -450,12 +450,16 @@ def rhsm_url
         end
       end
 
+      def api_url
+        URI("https://#{URI.parse(url).host}:8443")
```
Comment (ekohl) [inline app/models/katello/concerns/smart_proxy_extensions.rb:454]: This isn't quite correct. It only works on content proxies. On a Foreman + Smart Proxy this doesn't exist. So for the common case where there's only a single server you're breaking things. There is no way to detect if this is present or not. It can even be changed via an installer argument.

Also note that in the future we want to move this service away. Then Foreman Proxy can run on port 8443, just as a vanilla Foreman.

This is precisely why we introduced the registration module in Smart Proxy: it's a proper service we can rely on.

Also why I'm hammering on that Foreman -> Client -> Foreman is such a bad idea.

**Review by ekohl — CHANGES_REQUESTED** [review]
At this point I'd strongly urge you to split this PR into two: one with a REX job that only applies the current content source on a host. IMHO that's uncontroversial and can be easily merged.

Then we can talk about the IMHO controversial part of changing host properties via REX.

File: app/views/foreman/job_templates/change_content_source.erb (hunk 3 above)
Comment (stejskalleos) [reply app/views/foreman/job_templates/change_content_source.erb:16]: > Then I'd recommend to extracted it to a snippet that is properly reusable.

But we want to have it as job, so users can make changes on the hosts after they changed content source in UI. We can't run snippets as job right?

File: app/models/katello/concerns/smart_proxy_extensions.rb (hunk 4 above)
Comment (stejskalleos) [reply app/models/katello/concerns/smart_proxy_extensions.rb:454]: Hmm, so how to solve the problem with calling Foreman API here: https://github.com/Katello/katello/pull/9756/files#diff-dc69411f7cfa18aebde7d457a2d44098317c9d33755139b130ca5c8bd9fd88eaR42

Should we call only Foreman API with the risk that it is not accessible from the host?

**Comment by stejskalleos** [comment]
Split into two PRs: https://github.com/Katello/katello/pull/9780 and this one

File: app/views/foreman/job_templates/change_content_source.erb (hunk 3 above)
Comment (ekohl) [reply app/views/foreman/job_templates/change_content_source.erb]: I was suggesting to extract the actual logic to a snippet and then create two job templates.

File: app/views/foreman/job_templates/change_content_source_and_host.erb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    smart_proxy = input_resource('smart_proxy')
+-%>
+
+<%= render_template('Change content source', { smart_proxy: smart_proxy.id }) %>
+
+# Update host
+curl --silent --show-error --fail --output /dev/null \
+   --request PUT --cacert $SSL_CA_CERT \
+   --user <%= input('api_user') %>:<%= input('api_user_password') %> \
+   "https://<%= rhsm_url.host %>:8443/api/hosts/<%= @host.id %>" \
```
Comment (evgeni) [inline app/views/foreman/job_templates/change_content_source_and_host.erb:42]: this wont work if `rhsm_url` points at Foreman (and not a Proxy)

Comment (stejskalleos) [reply app/views/foreman/job_templates/change_content_source_and_host.erb:42]: Hmm, good point. But after seeing https://github.com/Katello/katello/pull/9756#discussion_r746658820, the only way we can do this is just call to the Foreman directly. We can't use proxy

Comment (evgeni) [reply app/views/foreman/job_templates/change_content_source_and_host.erb:42]: shouldn't `rhsm_url.port` point at the right port?!

**Comment by jturel** [comment]
Looks like this needs a rebase now. Please let me know when it's ready for final review

**Comment by jturel** [comment]
@stejskalleos is this no longer needed then?

**Comment by stejskalleos** [comment]
@jturel The feature is still in plan but I will post it in the new PR.

**Comment by jturel** [comment]
@stejskalleos ok. please use a new redmine for that one since the job template is going into 4.3 and that's going to be GA here soon (few days time)

**Comment by stejskalleos** [comment]
Follow PR: https://github.com/Katello/katello/pull/9873
